WIP: Update maintainers guide (#45)
laminas/technical-steering-committee: MAINTAINERS.md

This guide is intended for *maintainers* — anybody with commit access to one or more laminas Framework
repositories.

We use [automatic relases](https://github.com/laminas/automatic-releases) for
managing the various laminas Framework repositories. At a glance, this means:

- two or more minor version branches of the current major release
  - the highest minor version branch must not be released (i.e. does not have matching version tags like `1.0.0` 

>     for `1.0.x`).

Reviewer: This part isn't entirely true or necessary. There can be a single branch with no releases existing (e.g., a 0.1.x branch), or where the migration encountered a `develop` branch that had no changes from the `master` branch, and we only published the release branch associated with the most recent release.

AFTER a release has been made using the automatic-releases workflow, there will always be a minimum of two release branches. But BEFORE, there may only be one.

  - all other minor version branches must be released (i.e. have one or more version tags).

> - only unrelased branches may receive *new features*.

Reviewer: ```suggestion
- only unreleased branches may receive *new features*.
```

- released branches must only receive bugfixes or security fixes.
- releases (tags, next minor branch, github releases) will be performed automatically
  by the automatic releases workflow. 

Maintainers can choose to release new maintenance releases with each new patch, or accumulate
patches until a significant number of changes are in place. Security fixes, however, must be
released immediately and coordinated with the TSC to ensure an advisory is made, a

> CVE is created, and that the fix is backported to the Long Term Support releases.

Reviewer: This part changes.

We don't have LTS anymore.

We have "security windows". Essentially, in the case of a security advisory and associated release(s), we make the patch to the earliest release branch we still support. From there, automatic-releases takes over, and does merge-up releases to each next latest release branch until it hits the latest. Each of those releases along the way will be listed in the security advisory.

This guide is intended for *maintainers* — anybody with commit access to one or more laminas Framework
repositories.

> We use [automatic relases](https://github.com/laminas/automatic-releases) for

Reviewer: ```suggestion
We use [automatic releases](https://github.com/laminas/automatic-releases) for
```

> managing the various laminas Framework repositories. At a glance, this means:

Reviewer: Should be "Laminas Project", not "laminas Framework".

## Workflow for merging Pull Requests

Pull Requests should only be merged into the branches they were made against. When a PR contains a 
feature and was made against an already released branch, you should work with the contributor to target
it against the next release branch. The PR should have a Milestone matching the next release of the

> target branch (for example milestone `1.0.1` for branch `1.0.x`).

Reviewer: This workflow works pretty much the opposite now.

Since the default branch is the latest release branch, the more common scenario is that somebody will submit a bugfix against the latest release branch, when it should go against the previous. In those cases, you'll need to re-target the PR and help the user rebase or cherry-pick against the previous release branch.

To sum up:

> - If the PR is a bugfix, it should target the lowest affected release branch

Reviewer: I think we only want to target the lowest affected release branch for _security fixes_. For bugfixes, the latest release branch with a release created from it would be enough.

So, as an example, let's say we had a repo with the following release branches:

- 2.7.x
- 2.8.x
- 2.9.x

and tags:

- 2.7.0
- 2.7.1
- 2.7.2
- 2.8.0
- 2.8.1

We'd apply a bugfix against the 2.8.x branch in that case, and then end up issuing a 2.8.2 release.

> - If the PR is a new feature, merge to `develop` **only**.

Reviewer: ```suggestion
- If the PR is a new feature, merge to the latest existing release branch against which no tags have been created.
  If none exists, create a new one from the latest existing release branch using `git switch -c {X.Y+1.x} {X.Y.x}`.
```

    the author is the one most familiar with the changes they are introducing, they are the party
    most likely to resolve conflicts correctly.

## Workflow for merging Pull Requests

Pull Requests should only be merged into the correct release branches:

- If the PR is a bugfix, it should target one of the previous release branches.
  The only exception is when the package is not yet released and therefore has only 
  one release branch.
- If the PR is a new feature, merge it to the latest release branch **only**.

You may merge it by using the green "Merge" button in github. If you do so, use only the

> "Create a merge commit" option. Do **not** use "Squash and merge" or "Rebase and merge". 

Reviewer: I don't think we we ever discussed this in the TSC. What's the rationale for "create a merge commit" as the selected option? "Rebase and merge" leads to a more linear tree, which makes granular reversions of changes easier, while "Squash and merge" makes it easy to back out the entire changeset at once (though I'd argue squash and merge is _never_ a good idea because you lose granularlity). I've often used the `merge --ff` option when manually merging, which is similar to "Rebase and merge".

Maybe @Ocramius could weigh in?

Author: Actually this is just a suggestion since there is official statement here and I tried to base this on the previous gitflow strategy. But that all is a matter of taste and whatever is deemed to be important about the commit history. 

- A merge commit makes the origin of a change more obvious in the commit graph IMHO. 
- A squashed commit will consider the PR as a single commit which is sometimes useful to get rid of small adjustment commits during review rounds. It requires discipline to create small changesets though (which is always encouraged from a maintainers pov) - but this might drive off some contributors that are not familiar with this process.
- With a rebase and merge you'd loose the ability to track commits for its origin, but it produces a linear commit history.

I could also rephrase it to a "Pick whatever suits the individual situation for the PR" - which might be the best option here.

Alternatively you may clone the repository and merge the PR locally. If you decide to do so, 

> do not apply additional changes other than changelog entries to the merge.

Reviewer: This doesn't detail _how_ to merge. Should it be a standard merge, with a merge commit? Or a `merge --ff`? Or should we rebase and then `merge --ff`? (This discussion is related to my comments about using the button for merging. We need consistency, but also a good rationale for what we choose. And it needs to be the same whether using the button, or doing it at the CLI.)

Author: That's right much like above I actually don't know how to start here. At first do we want a permissive approach or restrictive approach. Both has its pros and cons. Permissive offers more flexibility, restrictive is more streamlined. Whatever the descition is: Here it should be clear to a Maintainer, what's good to go and what should not be done in either a permissive ("Don't do x, y, z - otherwise do as you think it suits best") or restrictive ("Only do x, y, z") way.

```

Each version gets a changelog entry in the project's `CHANGELOG.md` file. Not all changes need to be

> noted; things like bugfixes, coding standards fixes, continuous integration changes, or typo fixes do not need
> to be communicated. However, anything that falls under an addition, deprecation, removal
> MUST be noted. Please provide a succinct but *narrative* description for the change you merge.

Reviewer: ```suggestion
noted; things like coding standards fixes, continuous integration changes, or typo fixes do not need
to be communicated. However, any addition, deprecation, removal, or fix with user impact
SHOULD be noted. Please provide a succinct but *narrative* description for the change you merge,
detailing the change, its impact on the user, and what changes they may need to make to their code.
```

You may also ask the contributor to add these entries to get a feeling for what they're trying to accomplish.

## Performing a Release

> Releases are automatically performed by closing a milestone.

Reviewer: ```suggestion
Releases are automatically performed by closing a milestone. All issues and pull
requests in the milestone MUST be closed before closing the milestone; otherwise,
the release will not be issued.
```
